README.md
# esc - Effect Size Computation for Meta Analysis <img src="man/figures/logo.png" align="right" />

This is an R implementation of the web-based 'Practical Meta-Analysis Effect Size Calculator' from David B. Wilson. The original calculator can be found at http://www.campbellcollaboration.org/escalc/html/EffectSizeCalculator-Home.php.

Based on the input, the effect size can be returned as standardized mean difference (`d`), Hedges' `g`, correlation coefficient effect size `r` or Fisher's transformation `z`, odds ratio or log odds effect size.

### Return values

The return value of all functions has the same structure:

* The effect size, whether being `d`, `g`, `r`, (Cox) odds ratios or (Cox) logits, is always named `es`.
* The standard error of the effect size, `se`.
* The variance of the effect size, `var`.
* The lower and upper confidence limits `ci.lo` and `ci.hi`.
* The weight factor, based on the inverse-variance, `w`.
* The total sample size `totaln`.
* The effect size measure, `measure`, which is typically specified via the `es.type`-argument.
* Information on the effect-size conversion, `info`.
* A string with the study name, if the `study`-argument was specified in function calls.

#### Correlation Effect Size

If the correlation effect size `r` is computed, the transformed Fisher's z and their confidence intervals are also returned. The variance and standard error for the correlation effect size r are always based on Fisher's transformation.

#### Odds Ratio Effect Size

For odds ratios, the variance and standard error are always returned on the log-scale!

### S3 methods

The `esc` package offers the S3 methods `print` and `as.data.frame`

### Combining results into a single data frame

The `combine_esc` method is a convenient way to create pooled data frames of different effect size calculations, for further use. Here is an example of `combine_esc`, which returns a `data.frame` object.

```
e1 <- esc_2x2(grp1yes = 30, grp1no = 50, grp2yes = 40, grp2no = 45, study = "Study 1")
e2 <- esc_2x2(grp1yes = 30, grp1no = 50, grp2yes = 40, grp2no = 45, es.type = "or", study = "Study 2")
e3 <- esc_t(p = 0.03, grp1n = 100, grp2n = 150, study = "Study 3")
e4 <- esc_mean_sd(grp1m = 7, grp1sd = 2, grp1n = 50, grp2m = 9, 
                  grp2sd = 3, grp2n = 60, es.type = "logit", study = "Study 4")

combine_esc(e1, e2, e3, e4)

>     study         es    weight sample.size        se        var       ci.lo      ci.hi measure
> 1 Study 1 -0.3930426  9.944751         165 0.3171050 0.10055556 -1.01455689  0.2284717   logit
> 2 Study 2  0.6750000  9.944751         165 0.3171050 0.10055556  0.36256305  1.2566780      or
> 3 Study 3  0.2817789 59.433720         250 0.1297130 0.01682547  0.02754605  0.5360117       d
> 4 Study 4 -1.3981827  7.721145         110 0.3598812 0.12951447 -2.10353685 -0.6928285   logit
```


**esc** is still under development, i.e. not all effect size computation options are implemented yet. The remaining options will follow in further updates.

## Installation

### Latest development build

To install the latest development snapshot (see latest changes below), type following commands into the R console:

```r
library(devtools)
devtools::install_github("strengejacke/esc")
```

### Officiale, stable release

[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/esc)](https://cran.r-project.org/package=esc)
&#160;&#160;
[![downloads](http://cranlogs.r-pkg.org/badges/esc)](http://cranlogs.r-pkg.org/)
&#160;&#160;
[![total](http://cranlogs.r-pkg.org/badges/grand-total/esc)](http://cranlogs.r-pkg.org/)

To install the latest stable release from CRAN, type following command into the R console:

```r
install.packages("esc")
```

## Citation

In case you want / have to cite my package, please use `citation('esc')` for citation information.


NEWS
Version 0.3.2
------------------------------------------------------------------------------
Bug fixes:
* `combine_esc()` did not work for effect sizes computed with `esc_t()`, when the returned effect size was `r`.

Version 0.3.1
------------------------------------------------------------------------------
Bug fixes:
* `effect_sizes()` did not always properly extract study names and used numeric indices instead character values.

Version 0.3.0
------------------------------------------------------------------------------
New functions:
* `effect_sizes()` to generate a data frame with results of multiple effect size computations, based on data from another data frame.

Bug fixes:
* Fix control statements with condition with greater than one, which currently give a warning, but in future R versions may throw an error.


Version 0.2.0
------------------------------------------------------------------------------
New functions:
* `write_esc()` as a convenient wrapper to write results to an Excel csv-file.

Version 0.1.0
------------------------------------------------------------------------------
General:
* Initial release on CRAN.

R/esc.R
#' @title Generate effect size data frame from other data
#' @name effect_sizes
#'
#' @description This method computes any effect size from raw values from a data
#'              frame. Convenient method to compute multiple effect sizes at once,
#'              when the required information to calculate effects sizes are
#'              stored in a table (i.e. data frame).
#'
#' @param data A data frame with columns that contain the values that are passed
#'          to one of the \pkg{esc}-functions.
#' @param ... Named arguments. The name (left-hand side) is the name of one of
#'          \pkg{esc} functions' argument, the argument (right-hand side) is the
#'          name of the column in \code{data} that holds the data values.
#'          See 'Examples'.
#' @param fun Name of one of the \pkg{esc}-functions, as string, where arguments
#'          in \code{...} are passed to. May either be the full function name
#'          (like \code{"esc_t"} or \code{"esc_2x2"}) or the funcion name
#'          \emph{without} the suffix \code{"esc_"} (like \code{"t"} or \code{"2x2"}).
#' @inheritParams esc_beta
#'
#' @details This function rowwise iterates \code{data} and calls the function named
#'         in \code{fun} for the values taken from each row of \code{data}.
#'         The column names in \code{data} that contain the necessary values to compute
#'         the effect sizes should be passed as unquoted value for the arguments.
#'         The argument names should match those arguments for the esc-function
#'         that should be called from within \code{effect_sizes()}.
#'         \cr \cr
#'         Example: \cr
#'         If you want to compute effect sizes from chi-squared values, you
#'         would call \code{esc_chisq()}. This function name is used for the
#'         \code{fun}-argument: \code{fun = "esc_chisq"}. \code{esc_chisq()}
#'         requires one of \code{chisq} or \code{p} as arguments, and \code{totaln}.
#'         Now \code{data} must have columns with values for either \code{chisq}
#'         or \code{p}, and \code{effect_sizes()} automatically selects the
#'         first non-missing value from \code{data} (see 'Examples').
#'
#' @return A data frame with the effect sizes computed for all data from \code{data}.
#'
#' @examples
#' tmp <- data.frame(
#'   tvalue = c(3.3, 2.9, 2.3),
#'   n = c(250, 200, 210),
#'   studyname = c("Study 1", "Study 2", "Study 3")
#' )
#' effect_sizes(tmp, t = tvalue, totaln = n, study = studyname, fun = "esc_t")
#'
#' # missing effect size results are dropped,
#' # shorter function name, calls "esc_t()"
#' tmp <- data.frame(
#'   tvalue = c(3.3, 2.9, NA, 2.3),
#'   n = c(250, 200, 210, 210),
#'   studyname = c("Study 1", "Study 2", NA, "Study 4")
#' )
#' effect_sizes(tmp, t = tvalue, totaln = n, study = studyname, fun = "t")
#'
#'
#' tmp <- data.frame(
#'   coefficient = c(0.4, 0.2, 0.6),
#'   se = c(.15, .1, .2),
#'   treat = c(50, 60, 50),
#'   cntrl = c(45, 70, 40),
#'   author = c("Smith 2000", "Smith 2010 2", "Smith 2012")
#' )
#' effect_sizes(tmp, beta = coefficient, sdy = se, grp1n = treat, grp2n = cntrl,
#'     study = author, fun = "esc_beta", es.type = "or")
#'
#' # the "esc_chisq" function requires *either* the chisq-argument *or*
#' # the pval-argument. If at least one of these values is present,
#' # effect size can be calculated. You can specify both arguments,
#' # and the first non-missing required value from "data" is taken.
#' tmp <- data.frame(
#'   chisqquared = c(NA, NA, 3.3, NA, 2.9),
#'   pval = c(.003, .05, NA, .12, NA),
#'   n = c(250, 200, 210, 150, 180),
#'   studyname = c("Study 1", "Study 2", "Study 3", "Study 4", "Study 5")
#' )
#' effect_sizes(tmp, chisq = chisqquared, p = pval, totaln = n,
#'              study = studyname, fun = "esc_chisq")
#'
#' # if all required information are missing, data will be removed
#' tmp <- data.frame(
#'   chisqquared = c(NA, NA, 3.3, NA, NA),
#'   pval = c(.003, .05, NA, .12, NA),
#'   n = c(250, 200, 210, 150, 180),
#'   studyname = c("Study 1", "Study 2", "Study 3", "Study 4", "Study 5")
#' )
#' effect_sizes(tmp, chisq = chisqquared, p = pval, totaln = n,
#'              study = studyname, fun = "chisq")
#'
#'
#' @importFrom tibble as_tibble
#' @importFrom dplyr slice_
#' @importFrom purrr map_df
#' @export
effect_sizes <- function(data, ..., fun, es.type = c("d", "g", "or", "logit", "r", "cox.or", "cox.log")) {
  es.type <- match.arg(es.type)

  # check function name
  if (substr(fun, 1, 4) != "esc_") fun <- paste0("esc_", fun)

  # get arguments
  params <- match.call(expand.dots = FALSE)$`...`

  # and get columns names
  cols <- unname(unlist(lapply(params, function(x) as.character(x))))

  # make sure all numerics are numeric! and, only
  # return necessary data rows
  data <- suppressWarnings(
    purrr::map_df(data[, cols], function(x) {
      if (!is.factor(x) && !all(is.na(as.numeric(x))))
        as.numeric(x)
      else
        x
    })
  )

  results <- list()

  # iterate data frame rowwise
  for (i in 1:nrow(data)) {
    # get argument names
    argnames <- names(params)
    # get required data row
    datarow <- data[i, ]
    # copy values as list, for do.call
    callparams <- as.list(datarow)
    # give argument proper names
    names(callparams) <- argnames
    # call esc-method
    results[[length(results) + 1]] <-
      suppressWarnings(do.call(eval(fun), args = c(callparams, es.type = es.type)))
  }

  # returned results as combined tibble
  results <- suppressWarnings(purrr::map_df(results, function(x) tibble::as_tibble(x)))

  # do we have any missing effect sizes?
  if (anyNA(results$es) || any(is.infinite(results$es))) {
    # get studies with missing or Inf effect sizes
    mes <- which(is.na(results$es) | is.infinite(results$es))
    # warn user that some entries have been dropped
    warning(sprintf(
      "Row(s) %s in `data` had missing values and were removed.",
      paste0(mes, collapse = ", ")
    ),
    call. = F)
    # remove missings
    results <- dplyr::slice(results, -mes)
  }

  purrr::map_df(results, function(x) if (is.factor(x)) as.character(x) else x)
}